sparse/unit_test/Test_Sparse_spmv_bsr.hpp

```diff
@@ -14,563 +14,647 @@
+
+  const auto colids =
+      Kokkos::subview(rcs.get_ids(), Kokkos::pair{size_t(0), rcs.get_nnz()});
+  const auto vals =
+      Kokkos::subview(rcs.get_vals(), Kokkos::pair{size_t(0), rcs.get_nnz()});
+  Graph graph(colids, rcs.get_map());
+  Crs crs("crs", blockCols, vals, graph);
+
+  // expand to Bsr matrix
+  return KokkosSparse::Impl::expand_crs_to_bsr<Bsr>(crs, blockSize);
+}
 
-    y_vector_type ycrs("crs_product_result", nRow);
-    auto h_ycrs = Kokkos::create_mirror_view(ycrs);
-    for (lno_t ir = 0; ir < nRow; ++ir) h_ycrs(ir) = h_y0(ir);
-    Kokkos::deep_copy(ycrs, h_ycrs);
+/*! \brief reference SpMV is the KokkosSparse::spmv on the equivalent point
+ * matrix
+ */
+template <typename Alpha, typename Bsr, typename XVector, typename Beta,
+          typename YVector>
+void reference_spmv(const char *mode, const Alpha &alpha, const Bsr &a,
+                    const XVector &x, const Beta &beta, const YVector &y) {
+  using Crs = KokkosSparse::CrsMatrix<
+      typename Bsr::non_const_value_type, typename Bsr::non_const_ordinal_type,
+      typename Bsr::device_type, void, typename Bsr::non_const_size_type>;
+  const Crs crs = KokkosSparse::Impl::bsr_to_crs<Crs>(a);
+
+  KokkosSparse::spmv(mode, alpha, crs, x, beta, y);
+}
 
-    //
-    // Make reference computation with a CrsMatrix variable
-    //
-    KokkosKernels::Experimental::Controls controls;
-    // Use the native implementation since the CUDA 11.2.2 spmv implementation
-    // is not matching the bsr spmv test tolerance when OFFSET is int.
-    // See https://github.com/kokkos/kokkos-kernels/issues/1586
-#if defined(KOKKOSKERNELS_ENABLE_TPL_CUSPARSE) && (11200 <= CUSPARSE_VERSION)
-    controls.setParameter("algorithm", "native");
-#endif
-    KokkosSparse::spmv(controls, fOp, alpha, Acrs, xref, beta, ycrs);
-
-    y_vector_type ybsr("bsr_product_result", nRow);
-    auto h_ybsr = Kokkos::create_mirror_view(ybsr);
-    for (lno_t ir = 0; ir < nRow; ++ir) h_ybsr(ir) = h_y0(ir);
-    Kokkos::deep_copy(ybsr, h_ybsr);
-
-    // Create the BsrMatrix for the check test
-    KokkosSparse::Experimental::BsrMatrix<scalar_t, lno_t, device, void,
-                                          size_type>
-        Absr(Acrs, blockSize);
-
-    //
-    // Make computation with the BsrMatrix format
-    //
-    KokkosSparse::spmv(fOp, alpha, Absr, xref, beta, ybsr);
-
-    //
-    // Compare the two products
-    //
-    using KATS     = Kokkos::ArithTraits<scalar_t>;
-    using mag_type = typename KATS::mag_type;
-
-    const mag_type zero_mag = Kokkos::ArithTraits<mag_type>::zero();
-    mag_type error = zero_mag, maxNorm = zero_mag;
-
-    Kokkos::deep_copy(h_ycrs, ycrs);
-    Kokkos::deep_copy(h_ybsr, ybsr);
-    for (lno_t ir = 0; ir < nRow; ++ir) {
-      error   = std::max<mag_type>(error, KATS::abs(h_ycrs(ir) - h_ybsr(ir)));
-      maxNorm = std::max<mag_type>(maxNorm, KATS::abs(h_ycrs(ir)));
+template <typename Bsr>
```

Comment:
Could replace this with
``bsr_.blockDim() * KokkosSparse::Impl::graph_max_degree<device_t, Ordinal>(bsr_.graph.row_map)``